RavenDB-17241

redknightlois opened the pull request:
Support for comparison operations.

ayende commented on src/Corax/Queries/BinaryMatch.Boolean.cs:
```diff
@@ -0,0 +1,170 @@
+                }
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
this is called via a function pointer, I don't think it _can_ be inlined

redknightlois replied:
For now it can't. Don't know in the near future.

ayende commented on src/Corax/Queries/SortingMatch.cs:
```diff
@@ -292,6 +278,8 @@ public int Fill(Span<long> matches)
                 totalMatches = kIdx;
             }
+
+            QueryContext.MatchesPool.Return(matchesKeysHolder);
```
We intentionally avoid `try /finally`  here?

redknightlois replied:
Yes, there is actually no upside. Worst thing that can happen if you fail here --- which you shouldn't --- is increase allocations; but `try/finally` statements do disable many optimizations (afaik they are treated like `try/catch`).

ayende commented on src/Corax/Queries/UnaryMatch.Ordinal.cs:
```diff
@@ -0,0 +1,248 @@
+            var searcher = match._searcher;
+            var currentMatches = matches;
+            int totalResults = 0;
+            while (currentMatches.Length > matches.Length / 8)
```
Why do you need `/8` space here?

redknightlois replied:
I dont want to end on a situation where we keep looping and wasting resources for very small upside. One optimization I performed that had huge impact was avoiding to go to the end in `And` operations.

ayende replied:
Make sense, but in this case, we need to make sure that the size is big enough to do that at least once.
Basically, ensure that we don't have sizes ` < 8`

redknightlois replied:
Good point.

ayende commented on src/Corax/Queries/UnaryMatch.cs:
```diff
@@ -0,0 +1,72 @@
+        GreaterThanOrEqual,
+        LessThan,
+        LessThanOrEqual,
+        Equal,
+        NotEqual,
+        Not
```
Those aren't supported?

redknightlois replied:
`Not` and `NotEqual` haven't implemented them yet. Actually `Equal` already exist, different method, should be deleted.

ayende replied:
The reason this is interesting is that `negation` is an interesting problem to have. Lucene does that with exclusion on the other query clause.

redknightlois replied:
I would not know how to implement `Not` here. I pushed `Equals` and `NotEquals`

redknightlois committed:
RavenDB-17241 - Handling batches of different sizes properly.

redknightlois committed:
RavenDB-17241 - Between and NotBetween on content

redknightlois committed:
RavenDB-11275 - Instead of using stack space for buffers, use arrays.

redknightlois committed:
RavenDB-17241 - Equals and NotEquals on content.

redknightlois committed:
RavenDB-17241 - Operations support for GreaterThan & friends.

Supported operators:
- GreaterThan
- GreaterThanOrEqual
- LessThan
- LessThanOrEqual

redknightlois committed:
RavenDB-17241 - Support for Take requests

redknightlois commented:
OK. Unless we find anything else, this LGTM and could be merged. cc @ppekrol
